src/transaction/transaction_manager.cpp

```diff
@@ -99,23 +85,42 @@ timestamp_t TransactionManager::Commit(TransactionContext *const txn, transactio
   TERRIER_ASSERT(!txn->must_abort_,
                  "This txn was marked that it must abort. Set a breakpoint at TransactionContext::MustAbort() to see a "
                  "stack trace for when this flag is getting tripped.");
-  const timestamp_t result = txn->undo_buffer_.Empty() ? ReadOnlyCommitCriticalSection(txn, callback, callback_arg)
-                                                       : UpdatingCommitCriticalSection(txn, callback, callback_arg);
+  bool is_read_only = txn->undo_buffer_.Empty();
+  const timestamp_t result = is_read_only ? time_++ : UpdatingCommitCriticalSection(txn, callback, callback_arg);
   while (!txn->commit_actions_.empty()) {
     txn->commit_actions_.front()();
     txn->commit_actions_.pop_front();
   }
-
+  // Oldest active txn at the time of this commit
+  timestamp_t oldest_active_txn = INVALID_TXN_TIMESTAMP;
   {
     // In a critical section, remove this transaction from the table of running transactions
     common::SpinLatch::ScopedSpinLatch guard(&curr_running_txns_latch_);
     const timestamp_t start_time = txn->StartTime();
     const size_t ret UNUSED_ATTRIBUTE = curr_running_txns_.erase(start_time);
     TERRIER_ASSERT(ret == 1, "Committed transaction did not exist in global transactions table");
+
+    // If it's not a read-only txn and logging is enabled, we want to record the oldest active txn at the time we
+    // committed. This is used by recovery to serialize transactions. For read-only txns, its safe to pass off
+    // INVALID_TXN_TIMESTAMP, since they should not flush log records to disk.
+    if (!is_read_only && log_manager_ != LOGGING_DISABLED) {
+      const auto &oldest_txn = std::min_element(curr_running_txns_.cbegin(), curr_running_txns_.cend());
+      oldest_active_txn = (oldest_txn != curr_running_txns_.end()) ? *oldest_txn : NO_ACTIVE_TXN;
```

Comment:
I agree that computing it before we remove the txn is more conservative. I still believe the race condition exists, because the race condition we discussed has to do with the ordering of the logs, not the ordering of when we compute `oldest_active_txn`.